Document anti-patterns 📜 (#20)
carsdotcom/pacer: README.md

  <img src="assets/PACER.png" alt="Pacer" />
</object>

> # Pacer: An Elixir Library for Dependency Graph-Based Workflows With Robust Compile Time Safety & Guarantees

Author: I had a markdown formatter that went crazy here. I can revert it and only add the changes if refered.

## Overview

`Pacer.Workflow` is an abstraction designed for complex workflows where many interdependent data points need to be
stitched together to provide a final result, specifically workflows where each data point needs to be
loaded and/or calculated using discrete, application-specific logic.

See the moduledocs for `Pacer.Workflow` for detailed information about how to use Pacer and define your own
workflows, along with a detailed list of options and ideas that underlie Pacer.

## Installation

Add :pacer to the list of dependencies in `mix.exs`:

```elixir
[
  {:pacer, "~> 0.1"}
]
```

> ## Anti-patterns

Author: This is the only section I added.

I kind of like that it's a separate repo because it contains bad code and I don't want someone to blindly look at the Pacer code for examples on how to use it.

It also pulls in Benchee and has some pointless tests to run, which we might not want as part of the Pacer library.

That said, I'm open to pulling more complete examples into this repo.

Reviewer: I think it's totally fine to keep it as a separate repo. 

If we want to, we could eventually pull it in as a separate directory in this repo and call it something like `examples/anti-pattterns/` or something like that. 

OpenTelemetry Erlang does something like this [here](https://github.com/open-telemetry/opentelemetry-erlang/tree/main/samples).

Reviewer 2: Elixir [too](https://github.com/elixir-lang/elixir/tree/main/lib/elixir/pages/anti-patterns)!

- Using Pacer when your dependency graph is a line, or if you just have a

>   handfull of data points. You would be better off using a pipeline of function

Reviewer 2: ```suggestion
  handful of data points. You would be better off using a pipeline of function
```

Author: Thank you!

  calls. It would be easier to write, clearer to read, and faster to execute.

> - Not running your workflow in your tests. The main public interface of your
>   workflow is calling `Pacer.Workflow.execute/1` on it. Having a test file
>   for a Pacer workflow that does not do this, is like having any other module
>   where the main public interface is not exercised in the tests.

Reviewer 2: Maybe something like:

```suggestion

- Pacer does not run your workflow in when testing--the main public interface of your
  workflow is calls `Pacer.Workflow.execute/1`. Having a test file
  for a Pacer workflow that does not call `execute/1` is like having any other module
  where the main public interface is not exercised in the tests.
```

Author: Thanks for the feedback. I made a slightly different change. Please take a look.
